File: AI-Games/Assignment2/Videh/astar.cpp
//Heuristic values are given as input calculated using Manhattan distance
/*Sample input->
10   	//edges
10		//vertices
0 1 1
0 3 2
1 2 3
1 4 7
3 4 3		// vertices across edges along weights
4 6 7
6 7 4
7 9 2
7 8 3
8 5 5
5 4 3 4 3 3 2 1 2 0   //heuristic values
0 9 // start, stop points
*/
/*Corressponding maze->
	(1,1),(1,2),(1,3),(None)
	(2,1),(2,2),(None),(None)
	(None),(3,2),(3,3),(None)
	(None),(None),(4,3),(None)

	(1,1)->source
	(4,3)->goal
*/
/*
finally the path from the output->
0->3->4->6->7->9
OR
1,1->2,1->2,2->3,2->3,3->4,3
*/
#include<bits/stdc++.h>
using namespace std;

void astar(int v,vector<int>heuristic,vector<pair<int,int>>adj[],int source,int goal){
	priority_queue<pair<int,int>,vector<pair<int,int>>,greater<pair<int,int>>>pq;
	pq.emplace(make_pair(0,source));
	vector<int>cost_so_far(v,100000);  //initial high value
	vector<int>came_from(v,100000);	   //initial high value
	bool visited[v]; memset(visited,false,sizeof(visited));
	cost_so_far[source]=0;
	int current,new_cost,priority,neighbor;
	while(!pq.empty()){
		current=pq.top().second;
		pq.pop();
		visited[current]=true;
		if(current==goal)
			break;
		for(int i=0;i<adj[current].size();i++){
			neighbor=adj[current][i].first;
			//adj[current][i].first will give the neighbors of current vertex
			new_cost=cost_so_far[current] + adj[current][i].second;
			if((new_cost < cost_so_far[neighbor]) || visited[neighbor]==false){
				cost_so_far[neighbor]=new_cost;
				priority=new_cost+heuristic[neighbor];
				pq.emplace(make_pair(priority,neighbor));
				came_from[neighbor]=current;
			}
		}

	}
	for (int i = 0; i < came_from.size(); ++i)
	{
		cout<<came_from[i]<<" ";
	}
}

int main(){
	int e,v,u,w,source,weight,goal;
	cout<<"enter number of edges: ";
	cin>>e;
	cout<<"enter number of vertices: ";
	cin>>v;
	cout<<"tell the vertices between which there exist an edge and corressponding weights\n";
	
	vector<pair<int,int>>adj[v];

	for (int i = 0; i < e; ++i)
	{
		cin>>u>>w>>weight;
		adj[u].push_back(make_pair(w,weight));
		//adj[u].push_back(weight);
		adj[w].push_back(make_pair(u,weight));
		//adj[w].push_back(weight);
	}
	vector<int>heuristic;
	cout<<"enter heuristic values:\n";
	for (int i = 0; i < v; ++i)
	{
		int x;cin>>x;
		heuristic.push_back(x);
	}
	cout<<"enter the source node and goal node: ";
	cin>>source>>goal;
	cout<<endl<<"Adjacency list"<<endl;
	for (int i = 0; i < v; ++i)
	{
		cout<<i<<" -> ";
		for (int j = 0; j < adj[i].size(); ++j)
		{
			cout<<adj[i][j].first<<","<<adj[i][j].second<<"->";
		}
		cout<<endl;
	}
	cout<<endl;

	astar(v,heuristic,adj,source,goal);	
}

File: AI-Games/Assignment 2/Shruthan/astar.py
# -*- coding: utf-8 -*-
"""
Created on Sat May 25 23:10:17 2019

@author: shruthan
"""

import heapq

# Class to create the square grid for the maze    
class SquareGrid:
    def __init__(self, width, height):
        """
        Sets the height and width of the maze and also stores the walls in a list.
        """
        self.width = width
        self.height = height
        self.walls = []
    
    def in_bounds(self, id):
        """
        Checks if a given point in within the maze. Returns True or False
        """
        (x, y) = id
        return 0 <= x < self.width and 0 <= y < self.height
    
    def passable(self, id):
        """
        passable(vertex), Returns True if the vertex is not a wall and False otherwise
        """
        return id not in self.walls
    
    def neighbors(self, id):
        """
        Returns all neighbors(diagonal moves not considered) of a cell in the grif
        """
        (x, y) = id
        results = [(x+1, y), (x, y-1), (x-1, y), (x, y+1)]
        if (x + y) % 2 == 0: results.reverse()
        #To ensure neighbors in grid are within the maze.
        results = filter(self.in_bounds, results) 
        #To ensure neighbors are not walls.
        results = filter(self.passable, results) 
        return results

# Class to create a graph and store weights of edges
class GridWithWeights(SquareGrid):
    """
    Inherits from the SquareGrid class.
    """
    def __init__(self, width, height):
        """
        Initialises height and weights and creates a dictionary to store the weights.
        """
        # Calls __init__  of base class ie., SquareGrid with parameters height and weight.
        super().__init__(width, height) 
        self.weights = {}
    
    def cost(self, from_node, to_node):
        """
        Gets the value corresponding to the key to_node. Returns 1 if key does not exist
        """
        return self.weights.get(to_node, 1) 





def reconstruct_path(came_from, start, goal):
    """
    Returns the path from start to goal by reconstructing the order of visiting vertices.
    Takes in came_from as an argument which specifies the parent of each vertex.
    """
    current = goal
    path = []
    while current != start:
        path.append(current)
        current = came_from[current]
    path.append(start)
    path.reverse() # To give the order of travel from the begining.
    
    return path

def heuristic(a, b):
    """
    Returns the Manhattan distance between the two points given as a and b.
    """
    (x1, y1) = a
    (x2, y2) = b
    return abs(x1 - x2) + abs(y1 - y2)



def a_star_search(graph, start, goal):
    """
    Implements the A* search algorithms given the graph (which is an instance of the class GridWithWeights), start and end points.
    """
    open_list = list()
    heapq.heappush(open_list, (0, start))
    closed_list = {}
    cost_so_far = {}
    closed_list[start] = None
    cost_so_far[start] = 0
    
    
    while open_list:
        current = heapq.heappop(open_list)[1]
        
        #If goal already reached.
        if current == goal:
            path = (reconstruct_path(closed_list, start, goal))
            break
            
            
        
        for neighbor in graph.neighbors(current):
            new_cost = cost_so_far[current] + graph.cost(current, neighbor)
            if neighbor not in cost_so_far or new_cost < cost_so_far[neighbor]:
                cost_so_far[neighbor] = new_cost
                priority = new_cost + heuristic(goal, neighbor)
                heapq.heappush(open_list, (priority, neighbor))
                closed_list[neighbor] = current
    return path
   
    






def main():
    
    h = int(input("Enter the height of the maze:"))
    w = int(input("Enter the width of the maze:"))
    
    graph = GridWithWeights(w,h)
    
    w = int(input("Enter the number of walls:"))
    print("Enter the walls as a,b :")
    for i in range(w):
        a, b = input().split(",")
        a = int(a)
        b = int(b)
        graph.walls.append((a, b))
    n = int(input("Enter number of vertices: "))
    print("Enter the vertices and weights as vertex_x,vertex_y,weight : ")
    for i in range(n):
    
        x, y, w = input().split(",")
        x = int(x)
        y = int(y)
        w = int(w)
        graph.weights[(x,y)] = w
        
    print("Enter the starting vertex: ")
    x1, y1 = input().split(",")
    x1 = int(x1)
    y1 = int(y1)
    start = (x1, y1)
    
    print("Enter the goal vertex: ")
    x2, y2 = input().split(",")
    x2 = int(x2)
    y2 = int(y2)
    goal = (x2, y2)
    
    
    
    a_star_search(graph, start, goal)
    
    
    
    
    

# Sample input from one of the resources. Can be used instead of the main function
graph = GridWithWeights(10, 10)
graph.walls = [(1, 7), (1, 8), (2, 7), (2, 8), (3, 7), (3, 8)]
graph.weights = {loc: 5 for loc in [(3, 4), (3, 5), (4, 1), (4, 2),
                                       (4, 3), (4, 4), (4, 5), (4, 6), 
                                       (4, 7), (4, 8), (5, 1), (5, 2),
                                       (5, 3), (5, 4), (5, 5), (5, 6), 
                                       (5, 7), (5, 8), (6, 2), (6, 3), 
                                       (6, 4), (6, 5), (6, 6), (6, 7), 
                                       (7, 3), (7, 4), (7, 5)]}
#Example search
a_star_search(graph, (0,0), (7,4))






